### PR TITLE
RFC: Over Window on watermark

### PR DESCRIPTION
**Pull request by TennyZhuang**
Signed-off-by: TennyZhuang <zty0826@gmail.com>

### CODE REVIEW

**Comment by TennyZhuang** [comment]
![https://viewer.diagrams.net/?tags=%7B%7D&highlight=0000ff&edit=_blank&layers=1&nav=1&title=over-window-multiple.drawio#R7Vlbs5owEP41zLQP7UAQ1FcvvU07nY7t6DlvKeRAeiChIQj21zdAEBC11DmnwY4vmv2yS5Ldb5eNauY8zN4yGPmfqIsCDehuppkLDQBDNybiK0d2JWKbZgl4DLtSqQZW%2BBeqLCWaYBfFLUVOacBx1AYdSghyeAuDjNG0rfZAg%2FaqEfRQB1g5MOiia%2Bxyv0QnYFzj7xD2%2FGplw56WMyGslOVJYh%2B6NG1A5lIz54xSXo7CbI6C3HmVX0q7Nydm9xtjiPA%2BBndb%2Bz6ZrR%2FeT6zPm%2FDrz82XafxKRifmu%2BrAyBXnlyJl3KceJTBY1uiM0YS4KH%2BqLqRa5yOlkQANAf5AnO9kMGHCqYB8HgZyFmWYbxrju%2FxRry0pLTL55ELYVQLhbLdpCg2rXKzNCqmyK8%2BXH%2Bqk2yQU04Q56IyvgKQfZB7iZ%2FSsfXBFViAaIrEfYcdQADnetvcBJT29vV4dQTGQQfyLgE5vAX3ygNoqAyo3uYVBIlf6FsWcIRhqwA7EtmffmRh5%2BUizZtGjUBIf1qJLhCAQVTMPeOpjjlYRLLyTisLdDqdcETGOsvNO7jpFGoxk0ZNV36yKYFrXUKPC%2FEb9nOjP5cauH1VkyhOy1%2BrJ3lOR%2Bjf0tTpuX2aOD4lwwDH6uo9CudDTI0M1iQ9ZvJfVsbjrzitnsd2XxSOVLLYvZzEYGotHumoWG%2FatSemdIEZ16%2FlThhhK%2B85qm40UWWOSX3iOdimwaFOUV3jLHlqFN8a33OifG%2BZ15IZ5QW4of28c5sYA3hvT%2F637Mfo28eXPaMoI3G07P1BMjtIXE4KEPO9OOpTki7yQV9SXqgk%2Btod2Sa1%2BLG34eUFTMvzr%2FhQMzZWgW3OvvFaA3o2g0gs%2F6DaCK%2BG4o%2ByN63tSrLwX3L%2FfBtMLgu6189o5POrLYVMph0eXcVh5z3bI4efs2YRY%2F3lVzDX%2BAjSXvwE%3D](https://user-images.githubusercontent.com/9161438/198935912-7771b230-736f-4b9e-8ded-0aa8073d2bee.png)

**Comment by st1page** [comment]
> ![https://viewer.diagrams.net/?tags=%7B%7D&highlight=0000ff&edit=_blank&layers=1&nav=1&title=over-window-multiple.drawio#R7Vlbs5owEP41zLQP7UAQ1FcvvU07nY7t6DlvKeRAeiChIQj21zdAEBC11DmnwY4vmv2yS5Ldb5eNauY8zN4yGPmfqIsCDehuppkLDQBDNybiK0d2JWKbZgl4DLtSqQZW%2BBeqLCWaYBfFLUVOacBx1AYdSghyeAuDjNG0rfZAg%2FaqEfRQB1g5MOiia%2Bxyv0QnYFzj7xD2%2FGplw56WMyGslOVJYh%2B6NG1A5lIz54xSXo7CbI6C3HmVX0q7Nydm9xtjiPA%2BBndb%2Bz6ZrR%2FeT6zPm%2FDrz82XafxKRifmu%2BrAyBXnlyJl3KceJTBY1uiM0YS4KH%2BqLqRa5yOlkQANAf5AnO9kMGHCqYB8HgZyFmWYbxrju%2FxRry0pLTL55ELYVQLhbLdpCg2rXKzNCqmyK8%2BXH%2Bqk2yQU04Q56IyvgKQfZB7iZ%2FSsfXBFViAaIrEfYcdQADnetvcBJT29vV4dQTGQQfyLgE5vAX3ygNoqAyo3uYVBIlf6FsWcIRhqwA7EtmffmRh5%2BUizZtGjUBIf1qJLhCAQVTMPeOpjjlYRLLyTisLdDqdcETGOsvNO7jpFGoxk0ZNV36yKYFrXUKPC%2FEb9nOjP5cauH1VkyhOy1%2BrJ3lOR%2Bjf0tTpuX2aOD4lwwDH6uo9CudDTI0M1iQ9ZvJfVsbjrzitnsd2XxSOVLLYvZzEYGotHumoWG%2FatSemdIEZ16%2FlThhhK%2B85qm40UWWOSX3iOdimwaFOUV3jLHlqFN8a33OifG%2BZ15IZ5QW4of28c5sYA3hvT%2F637Mfo28eXPaMoI3G07P1BMjtIXE4KEPO9OOpTki7yQV9SXqgk%2Btod2Sa1%2BLG34eUFTMvzr%2FhQMzZWgW3OvvFaA3o2g0gs%2F6DaCK%2BG4o%2ByN63tSrLwX3L%2FfBtMLgu6189o5POrLYVMph0eXcVh5z3bI4efs2YRY%2F3lVzDX%2BAjSXvwE%3D](https://user-images.githubusercontent.com/9161438/198935912-7771b230-736f-4b9e-8ded-0aa8073d2bee.png)

we need another exchange before the join

**Comment by TennyZhuang** [comment]
> we need another exchange before the join

![https://viewer.diagrams.net/?tags=%7B%7D&highlight=0000ff&edit=_blank&layers=1&nav=1#R7VrbcpswEP0aZtqHdAwYgl99adJO2jTjduz0TTEy0IBEhWzjfH2EkQxY2IPdNMiZvNja1ep29mjZxdbMQZReERD737ALQ83ouKlmDjXD0Du6w74yzTrX2KaZKzwSuNyoUIyDJyhGcu0icGFSMaQYhzSIq8oZRgjOaEUHCMGrqtkch9VVY%2BBBSTGegVDWTgKX%2BrnWMS4L%2FTUMPF%2BsrNu9vCcCwpifJPGBi1cllTnSzAHBmOatKB3AMANP4JKP%2B7ynd7sxAhFtMuB%2Baf9e9CfzL451O41%2B%2Fp3e9ZIL7p2ErsWBocvOz0VMqI89jEA4KrR9ghfIhdmsHSYVNjcYx0ypM%2BUfSOmaOxMsKGYqn0Yh74VpQKel9n021SeLS8OUz7wR1kJAlKynZaE0KhOLYRtJjJNR4sAleEFm8AA0BmcbIB6kB%2Bys3C7DrbQA98EVxBFk%2B2EGBIaABssqrwCnp7e1KzzIGtyJRzi09%2B7Qf3WorZRD%2Ba6XIFzwlX7FCSUQRJphh%2Bwc%2FQfCWl7W0qx%2B%2FMiM2Ic1lIkQhixqZg5f%2BQGF4xhs4FqxwF11516Il5BQmB4Ehfc6POjxqG%2BKILgqYqgudH4pfopxLw%2BjjGMbN%2BV09loN2SsmVIS%2BlgT7KJ35ALHN1dHXfWTGG7tOrLdNYqNbZfFWbo%2FFMpznxWK7KYu7SrHYPp3Fhmos7nZekcU3%2Fe%2FX9vRpTu7g9c3tw4%2BJ5U8vLhUi8Q7BTmO1LiqTEq1rT263SWKxyxLukwBlNUltIgE2mUTrQdiyWwzCtU7svTn6mg3p67RKX%2FME%2BrYefXfp%2B6rRtx5Hlfh7El2bpsLi3ZMiSYQuJ29fcYBq6RsgBJk8kDtnGGWLfOCF3se2Ce4oV%2BoJt5dwHuIVUr9o1jtd1bA05KB7XsHCqEnNDmZHigQLQ07Wxgy4WvomRbmRtJ6vbR9w6hTNcvV2ZhzuNuWwqRaHu6dxuPWkbZfD7ZfM4uewt%2F5e%2F9h6Zn%2BV3eS6WK90Ow7t8oh3Si%2BUo0j3oAblvVfj0lCtHBcP7ver0eBqOGdxNeQE%2Fhyvxv98ajCx%2BOfApq%2F0%2Fwtz9Aw%3D](https://user-images.githubusercontent.com/9161438/198949407-89556303-b5de-49c7-9db7-a8f7af8ee6b8.png)

File: rfcs/0008-over-window.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+# Over Window
+
+## Summary
+
+We will introduce the Over Window (OverWindow) to RisingWave in the RFC.
+
+Another well-known name of OverWindow is [Window Function][window_function_wiki]. Unfortunately, the name is highly ambiguous in streaming.
+
+Flink named the feature Over Aggregation][over_aggregation_flink], but we think Over Window can describe the part more concisely.
```
Comment (fuyufjh) [inline rfcs/0008-over-window.md:16]: Syntax error?

File: rfcs/0008-over-window.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+The output rows count of OverWindow is exact as the input rows, as is the primary key.
+
+### Distribution key
+
+In most window definitions, one partition key is specified, and the data should be exchanged by the partition key.
+
+### BatchOverWindow
+
+The implementation is trivial, the only thing we need to concern is that the batch aggregator also needs to support `retract`, which is much more similar to the stream aggregator.
```
Comment (liurenjie1024) [inline rfcs/0008-over-window.md:78]: I'm afraid this is not trivial. If supporting `retract` is not enough for holistic functions, such as `count distinct` and `min/max`

**Comment by liurenjie1024** [comment]
I think what's missing in this doc is clarify the aggregation functions we need to support. Per pg's doc, all aggregations functions are supported in window function. If we only provide the most naive implementation for now, e.g. O(n^2) time complexity for all aggregations functions, I think streaming and batch can share aggregation functions.

File: rfcs/0008-over-window.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+The design highly depends on [the WatermarkFilter and StreamSort operator](https://github.com/risingwavelabs/rfcs/pull/1).
+
+If the input stream is ordered, then the implementation is trivial.
+
+There are some limitations in stream aggregation:
+
+* If the sort key is specified, then a watermark must exist on the column.
+* If no sort key is specified, then the result is undefined but may be helpful; it's roughly the same as the proctime in Flink.
+* `FOLLOWING` is **banned; the max value of `frame_start` and `frame_end` is `CURRENT ROW`.
```
Comment (fuyufjh) [inline rfcs/0008-over-window.md:90]: I think this restriction can also be applied to batch. In another words, it should be placed under the "Definitions" section.

Comment (fuyufjh) [reply rfcs/0008-over-window.md:90]: By the way, as mentioned by @liurenjie1024 , we need a more clear scope of what to support and what not to, such as the aggregation functions, the `ORDER`-able functions, the kind of frame_clause, etc...

File: rfcs/0008-over-window.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+* Window functions that peek the future rows are **banned**, e.g., `lag`.
+
+### Multiple windows
```
Comment (fuyufjh) [inline rfcs/0008-over-window.md:93]: This section looks too sketchy for implementation. Perhaps need another doc?

**Comment by liurenjie1024** [comment]
Databricks SQL's syntax: https://docs.databricks.com/sql/language-manual/sql-ref-window-functions.html 
I think it's more concise.

File: rfcs/0008-over-window.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+We can support them in the frontend easily.
+
+Assume we have two windows:
+
+```plain
+WINDOW w1 AS (PARTITION BY p1 ORDER BY s1)
+WINDOW w2 AS (PARTITION BY p2 ORDER BY s3)
+```
+
+We can plan the query as two branches with a join to combine them:
```
Comment (BugenZhao) [inline rfcs/0008-over-window.md:106]: This sounds not that trivial under the current planner & fragmenter. 🤣

Comment (BugenZhao) [reply rfcs/0008-over-window.md:106]: May off-topic: based on a similar idea, we can also enable two-phase optimization as much as possible if we can split agg calls and join them back. 🤩

File: rfcs/0008-over-window.md (hunk 1 above)
Comment (TennyZhuang) [reply rfcs/0008-over-window.md:16]:
```suggestion
Flink named the feature [Over Aggregation][over_aggregation_flink], but we think Over Window can describe the part more concisely.
```

File: rfcs/0008-over-window.md (hunk 4 above)
Comment (TennyZhuang) [reply rfcs/0008-over-window.md:93]: Yes, but we can ban it first.

Comment (TennyZhuang) [reply rfcs/0008-over-window.md:93]: And IMO the image is clear enough.

File: rfcs/0008-over-window.md (hunk 5 above)
Comment (TennyZhuang) [reply rfcs/0008-over-window.md:106]: Yes, we need DAG planner for that.

File: rfcs/0008-over-window.md (hunk 4 above)
Comment (liurenjie1024) [reply rfcs/0008-over-window.md:93]: I agree that we should ban it for now.

File: rfcs/0008-over-window-on-watermark.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+In most window definitions, one partition key is specified, and the data should be exchanged by the partition key.
+
+### BatchOverWindow
+
+The implementation is trivial.
+
+### StreamOverWindow
+
+The design highly depends on [the WatermarkFilter and StreamSort operator](https://github.com/risingwavelabs/rfcs/pull/1).
```
Comment (xxchan) [inline rfcs/0008-over-window-on-watermark.md:101]:
```suggestion
The design highly depends on [the WatermarkFilter and StreamSort operator](https://github.com/risingwavelabs/rfcs/pull/2).
```

**Review by fuyufjh — APPROVED** [review]
Time to merge it?

**Pull request merged**